core/unit_test/TestMultiGPU.hpp

```diff
@@ -181,4 +181,54 @@ void test_scratch(TEST_EXECSPACE exec0, TEST_EXECSPACE exec1) {
   ASSERT_EQ(error0, 0);
   ASSERT_EQ(error1, 0);
 }
+
+// Test that stream synchronization behavior for various GPU APIs matches the
+// assumptions made in Kokkos for multi gpu support, namely, that any stream (no
+// matter which device it is created on) can be synced from any device.
+struct AccumulateTag {};
+struct CheckTag {};
+
+template <int N>
+struct StreamSyncFunctor {
+  int *value;
+  int *check;
+
+  KOKKOS_FUNCTION
+  void operator()(const AccumulateTag &, const int) const {
+    for (int i = 0; i < N; ++i) Kokkos::atomic_inc(value);
+  }
+
+  KOKKOS_FUNCTION
+  void operator()(const CheckTag &, const int) const { check[0] = value[0]; }
+};
+
+void test_stream_sync(TEST_EXECSPACE exec0, TEST_EXECSPACE exec1,
+                      const std::function<void()> &set_to_device_1) {
```

Comment:
I think you should pass a lambda that sets the device of a space instance passed as an argument, to make this test more readable.
```suggestion
void test_stream_sync(TEST_EXECSPACE exec0, TEST_EXECSPACE exec1,
                      const SetDevice& set_device) {
```
and call site:
```c++
test_stream_sync(execs[0], execs[1], [](const Kokkos::Cuda& exec) {
      KOKKOS_IMPL_CUDA_SAFE_CALL(cudaSetDevice(exec.cuda_device()));
    });
```
and
```c++
  // Force set internal device to dev1
  set_device(exec1);
```